src/Microsoft.AspNetCore.Authentication.OpenIdConnect/OpenIdConnectHandler.cs

```diff
@@ -55,26 +55,118 @@ public class OpenIdConnectHandler : RemoteAuthenticationHandler<OpenIdConnectOpt
 
         protected HttpClient Backchannel { get; private set; }
 
-        protected HtmlEncoder HtmlEncoder { get; private set; }
+        protected HtmlEncoder HtmlEncoder { get; }
 
-        public OpenIdConnectHandler(HttpClient backchannel, HtmlEncoder htmlEncoder)
+        protected string SignOutScheme { get; private set; }
+
+        public OpenIdConnectHandler(IOptions<AuthenticationOptions> sharedOptions, IOptionsFactory<OpenIdConnectOptions> options, ILoggerFactory logger, HtmlEncoder htmlEncoder, UrlEncoder encoder, IDataProtectionProvider dataProtection, ISystemClock clock)
+            : base(sharedOptions, options, dataProtection, logger, encoder, clock)
         {
-            Backchannel = backchannel;
             HtmlEncoder = htmlEncoder;
         }
 
-        public override async Task<bool> HandleRequestAsync()
+        /// <summary>
+        /// The handler calls methods on the events which give the application control at certain points where processing is occurring. 
+        /// If it is not provided a default instance is supplied which does nothing when the methods are called.
+        /// </summary>
+        protected new OpenIdConnectEvents Events
+        {
+            get { return (OpenIdConnectEvents)base.Events; }
+            set { base.Events = value; }
+        }
+
+        public override async Task InitializeAsync(AuthenticationScheme scheme, HttpContext context)
+        {
+            await base.InitializeAsync(scheme, context);
+            Events = Events ?? new OpenIdConnectEvents();
+
+            SignOutScheme = Options.SignOutScheme;
+            if (string.IsNullOrEmpty(SignOutScheme))
+            {
+                SignOutScheme = SignInScheme;
+            }
+
+            if (Options.StateDataFormat == null)
```

Comment:
Yeah so that's why I'm trying to move all the Options manipulations into 'validations' that occur as part of OptionsFactory.Get("schemeName")